src/node/desktop/src/main/select-r.ts

```diff
@@ -0,0 +1,265 @@
+/*
+ * select-r.ts
+ *
+ * Copyright (C) 2021 by RStudio, PBC
+ *
+ * Unless you have received this program directly from RStudio pursuant
+ * to the terms of a commercial license agreement with RStudio, then
+ * this program is licensed to you under the terms of version 3 of the
+ * GNU Affero General Public License. This program is distributed WITHOUT
+ * ANY EXPRESS OR IMPLIED WARRANTY, INCLUDING THOSE OF NON-INFRINGEMENT,
+ * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE. Please refer to the
+ * AGPL (http://www.gnu.org/licenses/agpl-3.0.txt) for more details.
+ *
+ */
+
+import { BrowserWindow, ipcMain } from 'electron';
+import { existsSync } from 'fs';
+import path from 'path';
+import { logger } from '../core/logger';
+import { findDefault32Bit, findDefault64Bit, findRInstallationsWin32 } from './detect-r';
+
+function buildHtmlContent(rInstalls: string[]): string {
+
+  const visitedInstallations: { [index: string]: boolean } = {};
+  const rListEntries: string[] = [];
+
+  // sort so that newer versions are shown first
+  rInstalls.sort((lhs, rhs) => {
+    return rhs.localeCompare(lhs);
+  });
+
+  // make the HTML for each installation entry
+  rInstalls.forEach(rInstall => {
+
+    // normalize separators, etc
+    rInstall = path.normalize(rInstall).replaceAll(/[/\\]+$/g, '');
+
+    // skip if we've already seen this
+    if (visitedInstallations[rInstall]) {
+      return;
+    }
+    visitedInstallations[rInstall] = true;
+
+    // check for 64 bit executable
+    const r64 = `${rInstall}/bin/x64/R.exe`;
+    if (existsSync(r64)) {
+      const html = `<option value="${r64}">[64-bit] ${rInstall}</option>`;
```

Comment:
We should probably HTML-escape pieces here as appropriate. Need to see if Node / Electron have safe HTML templating or something similar.

Reply:
(Repeat of a slack conversation, just for the "permanent record"): I ported our simple HTML escaper from C++ to node.js here: https://github.com/rstudio/rstudio/blob/f591ebfa21478b3c21aba9691445bc691202dc43/src/node/desktop/src/core/string-utils.ts#L23-L37

Could also look at the npm packages `escape-html` (not updated in 6 years, mind you) or `html-escaper` (more recently updated).

Reply:
I ended up restructuring things here so that the select HTML widget is populated via IPC when the modal dialog is initialized.